### PR TITLE
support longcat-flash-lite-sparse

### PR DESCRIPTION
**Pull request by aqni**
## Motivation

[LongCat-Flash-Lite-Sparse](https://huggingface.co/meituan-longcat/LongCat-Flash-Lite-Sparse) is a sparse variant of the LongCat-Flash model family released by Meituan. This PR adds and fixes support for serving this model in SGLang.

## Modifications

- **`longcat_flash.py`**
  - Fix the ordering of the zero-expert contribution: it is now added **after** the TP all-reduce on `final_hidden_states`, instead of before, so the zero-expert result is not incorrectly duplicated/scaled across TP ranks.
  - Add weight name remapping for `oe_embed_tokens{i}.weight` / `oe_embed_proj{i}.weight` checkpoint keys to the internal `model.ngram_embeddings.embedders.{i}.weight` / `model.ngram_embeddings.post_projs.{i}.weight` naming, so N-gram embedding checkpoints in this naming convention load correctly.
- **`n_gram_embedding.py`**
  - Cast `exclusive_oe_embedder_size_sums[-1]` to a Python `int` before passing it as `num_embeddings` to `VocabParallelEmbedding`, fixing a type error when the value is a tensor. (see https://github.com/sgl-project/sglang/pull/32737 )
- **`dsa_indexer.py`**
  - Disable DSA indexer fusion (`use_dsa_indexer_fusion`) when `index_k_norm_type == "rms"` (i.e., `k_norm` is `RMSNorm` instead of `LayerNorm`), since the fused kernel path does not support RMSNorm-based k-normalization.
  - Pad the query/weight head dimension via `_pad_heads_for_deep_gemm` before calling into `deepgemm_paged_mqa_logits_native` / `deepgemm_paged_mqa_logits_split`, fixing a shape mismatch in DeepGEMM's paged MQA logits kernel during decode/target-verify. （see https://github.com/sgl-project/sglang/pull/32738 ）

## Accuracy Tests

Verified correct generation output on LongCat-Flash-Lite-Sparse with the launch command below (TP8, NSA prefill backend `fa3`, bf16 KV cache) on H20-141G.

```bash
python3 -m sglang.launch_server \
  --trust-remote-code \
  --model-path $MODEL_PATH \
  --host 0.0.0.0 \
  --port 30000 \
  --tp 8 \
  --max-running-requests 64 \
  --mem-fraction-static 0.9 \
  --chunked-prefill-size 2048 \
  --nsa-prefill-backend fa3 \
  --kv-cache-dtype bfloat16
```

## Speed Tests and Profiling

Not applicable — this change does not affect the inference hot path or kernel computation

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #31992775799](https://github.com/sgl-project/sglang/actions/runs/31992775799)
Latest PR Test (Extra): :x: [Run #31992775624](https://github.com/sgl-project/sglang/actions/runs/31992775624)

### CODE REVIEW

**Comment by aqni** [comment]
@BBuf Could you please help review this PR when you have a moment?

File: python/sglang/srt/layers/attention/dsa/dsa_indexer.py [hunk 1]
```diff
@@ -1005,11 +1007,14 @@ def _get_topk_paged(
                 get_paged_mqa_logits_metadata_fn=deep_gemm.get_paged_mqa_logits_metadata,
             )
         elif use_dg_native:
+            q_fp8_padded, weights_padded, _ = self._pad_heads_for_deep_gemm(
```
Comment (BBuf) [inline python/sglang/srt/layers/attention/dsa/dsa_indexer.py:939]: **[P2] Preserve the native 16-head shape on SM100/SM120**

The pinned DeepGEMM implementation accepts 8/16/32/64 heads on `arch_major == 10` and `arch_major == 12`; only SM90 restricts this kernel to 32/64 heads. Since this model uses `index_n_heads=16`, calling `_pad_heads_for_deep_gemm()` unconditionally here expands both Q and weights to 32 heads on B200/SM120 as well, adding 16 zero heads to every paged decode/target-verify invocation.

Could we make this padding architecture-aware (or derive it from a per-architecture supported-head list), so SM90 pads 16 to 32 while SM100/SM120 keep the native 16-head input? Please also add H=16 correctness coverage for both paths and a Blackwell benchmark.

**Review by BBuf — CHANGES_REQUESTED** [review]
The correctness fixes in this PR generally look sound:

- Moving the replicated zero-expert contribution after the TP all-reduce avoids scaling it by `tp_size`.
- The N-gram key remapping matches the public checkpoint layout: `oe_embed_tokens{0..11}` and `oe_embed_proj{0..11}`.
- Disabling the current LayerNorm-specific indexer fusion for RMSNorm is appropriate.

Before merging, please address the architecture-specific paged-MQA padding noted inline and add regression coverage for the newly supported path:

1. H=16 numerical parity for both the split path (`next_n=1`) and native target-verify path (`next_n>=2`) on SM90, including the required padding.
2. H=16 coverage on SM100/SM120 confirming that native heads are preserved, plus a decode benchmark.
3. A loader test for the flattened N-gram checkpoint keys.
4. A TP regression check for the zero-expert contribution, for example TP1 versus TP8 numerical parity or an accuracy threshold.

The existing paged-MQA test only covers H=32/64 and calls the wrappers directly, while the existing LongCat Lite FP8 e2e test explicitly does not cover Sparse DSA. Also, the functional/GPU CI jobs have not run because the PR is behind the required base commit. Please rebase and rerun them.

The Speed Tests and Profiling section should also be updated because this change modifies paged MQA in the decode/target-verify hot path.

**Comment by aqni** [comment]
@BBuf Added a few tests/benchmarks alongside the fixes:

- `test_longcat_flash_lite_sparse.py`: an integration test (TP8, 8-GPU) guarding the zero-expert contribution / TP all-reduce ordering fix — verifies gsm8k accuracy stays at the fixed level rather than regressing to the buggy pre-all-reduce behavior.
- `test_longcat_flash_weight_loading.py`: a CPU unit test for LongcatFlashForCausalLM.load_weights, covering the flattened N-gram embedding checkpoint key remapping.
- `test_deepgemm_paged_mqa_logits.py` / `bench_deepgemm_paged_mqa_logits.py`: correctness test + benchmark for the DeepGEMM paged MQA-logits arch-aware head padding fix, covering num_heads ∈ {8, 16, 32, 64} and native/padded code paths on SM90/SM100/SM120.

⚠️ Only verified locally on 8xH20-3e so far. 

### Result

`bench_deepgemm_paged_mqa_logits.py` on H20-3e:
```
dsa-indexer-head-padding-performance:
   batch  seq_len_kv  next_n  H=8 padded (us)  H=16 padded (us)  H=32 native (us)  H=32 padded (us)  H=64 native (us)  H=64 padded (us)
0    8.0      2048.0     1.0         8.480000          8.512000          8.512000          8.480000          9.024000          9.120000
1    8.0      2048.0     2.0         9.088000          9.120000          9.088000          9.120000         10.272000         10.336000
2   32.0      2048.0     1.0        11.200000         11.200000         11.200000         11.232000         13.280000         13.280000
3   32.0      2048.0     2.0        13.504000         13.504000         13.504000         13.504000         18.144000         18.080000
4  128.0      8192.0     1.0        48.831999         48.735999         48.767999         48.735999         71.648002         71.840003
5  128.0      8192.0     2.0        70.784003         70.975997         71.199998         71.584001        133.376002        133.471996
```
